File: agentic_os/mission/resource_scheduler.py
"""P11 — Cross-Mission Resource Scheduling: fair-share + admission control across missions.

Per-mission concurrency (P0) is enough to *validate* the architecture; this plane matters once
multiple durable missions compete for genuinely scarce resources — operators, models, GPUs, API
quotas, monetary budget and (the one people forget) **human-approval capacity**. It arbitrates with
priorities, deadlines, organizational fair-share and starvation prevention.

Only SCARCE resources go in the pool; anything unpooled is treated as unlimited, so a mission is
never blocked on a resource nobody is rationing. Everything here is deterministic (a logical round
counter, no wall clock) so replay reproduces the same admission order.
"""
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field


@dataclass
class ResourceRequest:
    mission_id: str
    demand: dict                       # resource -> amount (only pooled keys are rationed)
    priority: int = 0                  # higher = more important
    deadline: float | None = None      # earliest-deadline-first; lower = sooner (None = no deadline)
    owner: str = "default"             # fair-share tenant (org / team / customer)


@dataclass
class _Waiting:
    req: ResourceRequest
    enqueued: int                      # logical submit sequence (FIFO tiebreak)
    rounds_waited: int = 0


class InfeasibleRequest(ValueError):
    """Demand exceeds total pool capacity for some resource — it could never be admitted."""


class CrossMissionScheduler:
    def __init__(self, pools: dict, *, aging_rate: float = 1.0):
        self.capacity = dict(pools)
        self.available = dict(pools)
        self.aging_rate = aging_rate
        self._q: list[_Waiting] = []
        self._seq = 0
        self._admitted: dict[str, dict] = {}                 # mission_id -> reserved demand
        self._served: dict[str, float] = defaultdict(float)  # owner -> cumulative units granted

    # ── queue ────────────────────────────────────────────────────────────────
    def submit(self, req: ResourceRequest) -> None:
        for r, amt in req.demand.items():
            if r in self.capacity and amt > self.capacity[r]:
                raise InfeasibleRequest(f"{req.mission_id} needs {amt} {r} > capacity {self.capacity[r]}")
        self._seq += 1
        self._q.append(_Waiting(req, self._seq))

    # ── admission control ─────────────────────────────────────────────────────
    def admit(self) -> list[str]:
        """One admission round. Age every waiter (starvation prevention), then greedily admit the
        best-ranked requests that still fit — fair-share first, then aged priority, then EDF, FIFO."""
        for w in self._q:
            w.rounds_waited += 1
        admitted: list[str] = []
        for w in sorted(self._q, key=self._rank):
            if self._fits(w.req.demand):
                self._reserve(w.req.demand)
                self._admitted[w.req.mission_id] = w.req.demand
                self._served[w.req.owner] += self._weight(w.req.demand)
                admitted.append(w.req.mission_id)
        if admitted:
            self._q = [w for w in self._q if w.req.mission_id not in self._admitted]
        return admitted

    def _rank(self, w: _Waiting):
        effective_priority = w.req.priority + self.aging_rate * w.rounds_waited
        return (
            self._served[w.req.owner],                                   # 1. least-served owner (fair share)
            -effective_priority,                                        # 2. higher aged priority
            w.req.deadline if w.req.deadline is not None else float("inf"),  # 3. earliest deadline
            w.enqueued,                                                 # 4. FIFO
        )

    def release(self, mission_id: str) -> None:
        demand = self._admitted.pop(mission_id, None)
        if demand:
            for r, amt in demand.items():
                if r in self.capacity:
                    self.available[r] = min(self.capacity[r], self.available.get(r, 0) + amt)

    # ── helpers ────────────────────────────────────────────────────────────────
    def _fits(self, demand: dict) -> bool:
        return all(self.available.get(r, float("inf")) >= amt for r, amt in demand.items())

    def _reserve(self, demand: dict) -> None:
        for r, amt in demand.items():
            if r in self.capacity:
                self.available[r] -= amt

    @staticmethod
    def _weight(demand: dict) -> float:
        return float(sum(demand.values())) or 1.0

    def snapshot(self) -> dict:
        return {"available": dict(self.available),
                "waiting": [w.req.mission_id for w in self._q],
                "admitted": list(self._admitted),
                "served": dict(self._served)}


def estimate_demand(plan) -> dict:
    """A mission's resource footprint from its compiled plan — monetary spend, human-approval
    capacity, execution slots and per-operator load. Only keys present in the pool are rationed."""
    g = plan.graph
    demand: dict = {
        "usd": round(sum(n.cost.usd for n in g.nodes), 4),
        "human_approval": sum(1 for n in g.nodes if n.approval_required),
        "slots": len(g.nodes),
    }
    for n in g.nodes:
        key = f"operator:{n.operator}"
        demand[key] = demand.get(key, 0) + 1
    return {k: v for k, v in demand.items() if v}

File: go/mission/resource_scheduler.go
package mission

// P11 — Cross-Mission Resource Scheduling: fair-share + admission control across missions competing
// for scarce resources (operators, models, GPUs, quotas, budget, human-approval capacity). Only
// SCARCE resources go in the pool; anything unpooled is unlimited. Deterministic (a logical round
// counter). Go port of agentic_os/mission/resource_scheduler.py.

import (
	"fmt"
	"math"
	"sort"
)

// ResourceRequest is one mission's demand on the shared pools.
type ResourceRequest struct {
	MissionID string
	Demand    map[string]float64 // resource → amount (only pooled keys are rationed)
	Priority  int
	Deadline  *float64 // earliest-deadline-first; nil = no deadline
	Owner     string   // fair-share tenant
}

type waiting struct {
	req          ResourceRequest
	enqueued     int
	roundsWaited int
}

// InfeasibleRequest: demand exceeds total pool capacity for some resource.
type InfeasibleRequest struct{ Msg string }

func (e *InfeasibleRequest) Error() string { return e.Msg }

type CrossMissionScheduler struct {
	capacity  map[string]float64
	available map[string]float64
	agingRate float64
	q         []*waiting
	seq       int
	admitted  map[string]map[string]float64
	served    map[string]float64
}

func NewCrossMissionScheduler(pools map[string]float64, agingRate float64) *CrossMissionScheduler {
	cap := map[string]float64{}
	avail := map[string]float64{}
	for k, v := range pools {
		cap[k] = v
		avail[k] = v
	}
	return &CrossMissionScheduler{capacity: cap, available: avail, agingRate: agingRate,
		admitted: map[string]map[string]float64{}, served: map[string]float64{}}
}

// Submit enqueues a request, rejecting an infeasible demand (> total capacity) up front.
func (s *CrossMissionScheduler) Submit(req ResourceRequest) error {
	for r, amt := range req.Demand {
		if c, pooled := s.capacity[r]; pooled && amt > c {
			return &InfeasibleRequest{Msg: fmt.Sprintf("%s needs %g %s > capacity %g", req.MissionID, amt, r, c)}
		}
	}
	s.seq++
	s.q = append(s.q, &waiting{req: req, enqueued: s.seq})
	return nil
}

// Admit runs one round: age every waiter, then greedily admit the best-ranked fitting requests
// (fair-share → aged priority → EDF → FIFO).
func (s *CrossMissionScheduler) Admit() []string {
	for _, w := range s.q {
		w.roundsWaited++
	}
	ranked := append([]*waiting{}, s.q...)
	sort.SliceStable(ranked, func(i, j int) bool { return s.rankLess(ranked[i], ranked[j]) })

	var admitted []string
	for _, w := range ranked {
		if s.fits(w.req.Demand) {
			s.reserve(w.req.Demand)
			s.admitted[w.req.MissionID] = w.req.Demand
			s.served[w.req.Owner] += weight(w.req.Demand)
			admitted = append(admitted, w.req.MissionID)
		}
	}
	if len(admitted) > 0 {
		var kept []*waiting
		for _, w := range s.q {
			if _, done := s.admitted[w.req.MissionID]; !done {
				kept = append(kept, w)
			}
		}
		s.q = kept
	}
	return admitted
}

func (s *CrossMissionScheduler) rankLess(a, b *waiting) bool {
	if s.served[a.req.Owner] != s.served[b.req.Owner] { // 1. least-served owner (fair share)
		return s.served[a.req.Owner] < s.served[b.req.Owner]
	}
	ea := float64(a.req.Priority) + s.agingRate*float64(a.roundsWaited)
	eb := float64(b.req.Priority) + s.agingRate*float64(b.roundsWaited)
	if ea != eb { // 2. higher aged priority
		return ea > eb
	}
	da, db := math.Inf(1), math.Inf(1) // 3. earliest deadline
	if a.req.Deadline != nil {
		da = *a.req.Deadline
	}
	if b.req.Deadline != nil {
		db = *b.req.Deadline
	}
	if da != db {
		return da < db
	}
	return a.enqueued < b.enqueued // 4. FIFO
}

func (s *CrossMissionScheduler) Release(missionID string) {
	demand, ok := s.admitted[missionID]
	if !ok {
		return
	}
	delete(s.admitted, missionID)
	for r, amt := range demand {
		if c, pooled := s.capacity[r]; pooled {
			s.available[r] = math.Min(c, s.available[r]+amt)
		}
	}
}

func (s *CrossMissionScheduler) fits(demand map[string]float64) bool {
	for r, amt := range demand {
		av, pooled := s.available[r]
		if pooled && av < amt {
			return false
		}
	}
	return true
}

func (s *CrossMissionScheduler) reserve(demand map[string]float64) {
	for r, amt := range demand {
		if _, pooled := s.capacity[r]; pooled {
			s.available[r] -= amt
		}
	}
}

func weight(demand map[string]float64) float64 {
	sum := 0.0
	for _, v := range demand {
		sum += v
	}
	if sum == 0 {
		return 1.0
	}
	return sum
}

func (s *CrossMissionScheduler) Snapshot() map[string]any {
	waitingIDs := make([]string, 0, len(s.q))
	for _, w := range s.q {
		waitingIDs = append(waitingIDs, w.req.MissionID)
	}
	admittedIDs := make([]string, 0, len(s.admitted))
	for id := range s.admitted {
		admittedIDs = append(admittedIDs, id)
	}
	avail := map[string]float64{}
	for k, v := range s.available {
		avail[k] = v
	}
	served := map[string]float64{}
	for k, v := range s.served {
		served[k] = v
	}
	return map[string]any{"available": avail, "waiting": waitingIDs, "admitted": admittedIDs, "served": served}
}

// EstimateDemand derives a mission's resource footprint from its compiled plan.
func EstimateDemand(plan *ExecutionPlan) map[string]float64 {
	g := plan.Graph
	usd := 0.0
	approvals := 0.0
	for i := range g.Nodes {
		usd += g.Nodes[i].Cost.USD
		if g.Nodes[i].ApprovalRequired {
			approvals++
		}
	}
	demand := map[string]float64{
		"usd": roundN(usd, 4), "human_approval": approvals, "slots": float64(len(g.Nodes)),
	}
	for i := range g.Nodes {
		demand["operator:"+g.Nodes[i].Operator]++
	}
	out := map[string]float64{}
	for k, v := range demand {
		if v != 0 {
			out[k] = v
		}
	}
	return out
}
